controllers/release/release_adapter.go

```diff
@@ -227,24 +228,40 @@ func (a *Adapter) EnsureSnapshotEnvironmentBindingIsCreated() (results.Operation
 // EnsureTargetContextIsSet is an operation that will ensure that the targetContext of the adapter is not nil. It will
 // set it to a context from the workspace in the ReleasePlan if one is defined or to its own context if not.
 func (a *Adapter) EnsureTargetContextIsSet() (results.OperationResult, error) {
-	if a.targetContext == nil {
+	if a.release.Status.Target == (kcp.NamespaceReference{}) {
+		patch := client.MergeFrom(a.release.DeepCopy())
 		releasePlan, err := a.getReleasePlan()
 		if err != nil {
-			patch := client.MergeFrom(a.release.DeepCopy())
 			a.release.MarkInvalid(v1alpha1.ReleaseReasonReleasePlanValidationError, err.Error())
 			return results.RequeueOnErrorOrStop(a.client.Status().Patch(a.context, a.release, patch))
 		}
 
-		if releasePlan.Spec.Target.Workspace == "" {
-			a.logger.Info("Running in a traditional kubernetes environment")
-			a.targetContext = a.context
-		} else {
+		if _, ok := logicalcluster.ClusterFromContext(a.context); ok {
 			a.logger.Info("Running in a KCP environment")
-			a.targetContext = logicalcluster.WithCluster(a.context, logicalcluster.New(releasePlan.Spec.Target.Workspace))
+
+			if releasePlan.Spec.Target.Workspace == "" {
+				a.release.MarkInvalid(v1alpha1.ReleaseReasonReleasePlanValidationError, "ReleasePlan Workspace is not set while running on KCP")
```

Comment:
Maybe it should say `ReleasePlan Workspace not set...`. Leaving this (and the one below) to @johnbieren.